PRC-4 draft (#5)
PaimaStudios/PRC: PRCS/prc-4.md

Instead of bridging, this standard allows trading the game assets on base chain via creating and filling sell orders in an orderbook DEX smart contract.

## Specification

Every PRC-4 compliant contract must implement the `IOrderbookDex` interface and events:

```solidity
/// @notice Facilitates base-chain trading of an asset that is living on a different app-chain.
/// @dev The contract should never hold any ETH itself.
interface IOrderbookDex is IERC165 {
    struct Order {
        uint256 assetAmount;
        uint256 price;
        address payable seller;
        bool active;
    }

    event OrderCreated(
        uint256 indexed orderId,
        address indexed seller,
        uint256 assetAmount,
        uint256 price
    );
    event OrderFilled(
        uint256 indexed orderId,
        address indexed seller,
        address indexed buyer,
        uint256 assetAmount,
        uint256 price

>     );

Reviewer: This isn't enough to support partially filled matches

2. The game chain monitors this contract using Paima Primitives and exposes an API to query its state.

The contract on the base chain has no way of knowing if somebody who makes a sell order really has that amount of game assets in their account. Rather,

1. When somebody creates a sell order: The game chain sees it and checks if the user has enough of the asset in their account.
    * If yes, the amount is locked so they cannot spend it in-game (unless they cancel the order).
    * If not, the order is marked as invalid (to differentiate it from the case where it doesn't know the order exists yet).
2. When somebody wants to buy, they specify the orders they wish to purchase by ID. 
    * The responsibility of querying the API of game chain to check the validity of the sell order is left up to the front-end providers, as well as the presentation of the possible sell orders.
3. The user then makes a smart contract call that fulfills all the orders and sends the right amount to the corresponding accounts in a single transaction (atomic).
4. The game chain monitors the successfully fulfilled orders on the base chain and transfers the assets between accounts.

> **Note:**

Reviewer: Although an increasing index works, it will make the UI slow to update (recall: we had the same problem in the inverse projection where if the inverse projection is base-initiated it takes a while for the game node to update)

In this case however, we can't easily rely on app-side initiation solely because new UTXO entries are created when an order is partially filled (which has a new index)

To solve it though, I think we can take a variation of what we did for the inverse projection system:

1. Initiate your sell order on Tarochi (by placing in your gold to trade). This will give you a wallet-specific counter (instead of a global counter like we have here)
2. The RPC endpoint we query to know is `https://${rpcBase}/inverseProjection/${purpose}/${chainIdentifier}/` just like in PRC3, but with `purpose` set to `orderbook`
3. Unlike the NFT case where we queried for the file `${address}/${userTokenId}`, we instead query for `${address}/${orderId}/${amount}`. The endpoint then returns something like `{ aboveInitial: true }` if the amount provided is above the initial tgold deposited by the user in step 1 (note above initial means when the initial utxo was created in this chain)

Note that this requires essentially 2 IDs that we need to work with:
1. An initial ID
2. A sub-ID for how far we are in a chain of partial orders

I'm open to thoughts about how we want to handle the combination of these ids. We have 2 options:
1. The `orderID` is a combination of `initial_id, sub_id>`. This means that the game just needs to track `orderIDs` and it can look up their balance just from that, and this is more faithful to the UTXO model. However, it means functions like `cancelSellOrder` or `fillSellOrders` may fail just because the subID changed even though the function call would have succeeded otherwise
3. The `orderID` is constant (subID is just a concept that isn't actually in the Solidity contract directly).

Then, like PRC3, we can create a function in this contract (one with the default RPC, one with the user RPC) that constructs the RPC URL from the Solidity contract.

With this configuration, we should end up where the Solidity contract provides an endpoint that always works right away to know if the order is real (has enough tgold). You simply check
1. if `active === true` in the Solidity contract
4. if `aboveInitial === true` from the RPC call

Author: With partial fills, why would we have to discard the old order and create a new order? I understand that you're thinking in terms of UTxO model, but is this really necessary? We can just have the order track the amount and have it change with partial fills.
Example: There is sell order for 1000 TGOLD. I submit a buy for 800 TGOLD but somebody else submits a buy of 600 TGOLD with higher gas fee, so by the time my transaction is being processed, 600 was already bought and so the sell order would only have 400 TGOLD left. I would much rather have my buy order fill for the 400 TGOLD (because it was a good price per token) than it to revert and me not getting anything and losing money on failed transaction gas fee.

Reviewer: > With partial fills, why would we have to discard the old order and create a new order

Yes, as I mentioned in my post above we can just do the option `orderID is constant`. I initially thought we would have to create new orders every time for the game to react based off of it, but I realized that wasn't necessary since the RPC endpoint I mentioned is enough for the game endpoint to return the right state without having to see the tx on Arbitrum

    function fillSellOrders(uint256[] memory orderIds) external payable;

    /// @notice Cancels the sell order specified by `orderId`, making it unfillable.
    /// @dev Reverts if the msg.sender is not the order's seller.
    /// MUST change the `active` parameter for the specified order to `false`.
    /// MUST emit `OrderCancelled` event.
    function cancelSellOrder(uint256 orderId) external;
}
```

## Rationale

It's not an AMM. In a typical AMM dex, there are liquidity pools and you trade against these pools. This is unsuitable for this use-case because a contract on base chain has no means of knowing the state of the game chain so the contract cannot automate market making. Instead, buyers have to be sending tokens directly to the sellers.

### The idea:

1. Smart contract on base chain is created to facilitate trading game asset. This contract allows the following:
    * Function to create a sell order. A sell order persists and has the following properties: 
      * `assetAmount` - amount of the game asset the seller is selling,
      * `price` - the price in native gas tokens of the base chain the seller is requesting,
      * `seller` - the address of the seller,

>       * `active` - signalizes if the sell order is active or not (meaning it has been cancelled or filled)

Reviewer: ```suggestion
      * `active` - signals whether or not the sell order is active (meaning it has been cancelled or filled)
```

    * Function to cancel a sell order.
    * Function to fill a sell order (in other words - buy). There is no "buy order" that persists, and rather the buy function directly transfers the value specified by price defined in the orders to the sellers. Buying marks all the sell orders it fulfills as inactive by changing the value of `active` flag to `false`.

> 2. The game chain monitors this contract using CDEs and exposes an API to query its state.

Reviewer: ```suggestion
2. The game chain monitors this contract using Paima Primitives and exposes an API to query its state.
```

> That is to say, the contract on the base chain has no way of knowing if somebody who makes a sell order really has that amount of game assets in their account. Rather,

Reviewer: ```suggestion
The contract on the base chain has no way of knowing if somebody who makes a sell order really has that amount of game assets in their account. Rather,
```

    * The responsibility of querying the API of game chain to check the validity of the sell order is left up to the front-end providers, as well as the presentation of the possible sell orders.
3. The user then makes a smart contract call that fulfills all the orders and sends the right amount to the corresponding accounts in a single transaction (atomic).
4. The game chain monitors the successfully fulfilled orders on the base chain and transfers the assets between accounts.

**Note:**
All actions (sell, buy, cancel) are done on the base chain (and not on the game chain). This allows the base chain to be the source of truth for the state of which orders are valid which avoids double-spends. That is to say, if you try and perform a buy order, you are guaranteed by the base chain itself that the order hasn't been fulfilled by anybody else yet.

## Reference Implementation

You can find all the contracts and interfaces in the Paima Engine codebase [here](https://github.com/PaimaStudios/paima-engine/blob/master/packages/contracts/evm-contracts/contracts/orderbook/).

## Considerations

### Avoiding many failed txs in this model

> The problem with this model is people will naturally all try to buy the order with the most favorable sell price. Chosen base chain should have fast blocks so collisions are not so likely if the UI updates fast enough, but if bots start trading good orders as soon as they appear it may cause a rush with many failed transactions.

Reviewer: ```suggestion
The problem with this model is people will naturally all try to buy the order with the most favorable sell price. The chosen base chain should have fast blocks so that collisions are not too likely if the UI updates fast enough, but if bots start trading good orders as soon as they appear it may cause a rush with many failed transactions.
```

**Option 1) A L3 for all games**

> However, the key observation is that solving this concurrency issue does not need any knowledge of the game itself (it doesn't matter if orders are valid or not from the dApp perspective. It just assumes buyers are not making bad purchases).

Reviewer: ```suggestion
The key observation is that solving this concurrency issue does not need any knowledge of the game itself (it doesn't matter if orders are valid or not from the dApp perspective. It just assumes buyers are not making bad purchases).
```

> This is an important note because it means we could have a decentralized L3 on top of the base chain whose goal it is to match orders properly. Notably, there is a model that provides exactly what we want with no tx fee if concurrent actions are attempted like this: the UTxO model. That is to say, implementing this system on top of Cardano (Aiken) or Fuel is actually much easier than the account model of EVM. However, e.g. Arbitrum users of course cannot be using Cardano, so this would most likely require running a Fuel L3 for Arbitrum (or maybe some ZK-ified UTxO platform if one exists) where the validators are decided by the Paima ecosystem token.

Reviewer: ```suggestion
This is an important note because it means we could have a decentralized L3 on top of the base chain whose goal it is to match orders properly. Notably, there is a model that provides exactly what we want with no tx fee if concurrent actions are attempted like this: the UTxO model. That is to say, implementing this system on top of Cardano (Aiken) or Fuel is actually much easier than the account model of EVM. However, e.g. Arbitrum users of course cannot be using Cardano, so this would most likely require running a Fuel L3 for Arbitrum (or maybe some ZK-ified UTxO platform) where the validators are decided by the Paima ecosystem token.
```

> It falls short on a few key points:

Reviewer: ```suggestion
However, this falls short on a few key points:
```

* It would require adding Fuel support to Paima Engine to properly monitor it.

> * It would require the Paima ecosystem to be released (not released yet).

Reviewer: ```suggestion
* It would require the Paima ecosystem token to be released (not released yet).
```

* It means we lose some composability with other Arbitrum dApps (unless Fuel adds a wrapped smart contract system like Milkomeda).
* Users may be hesitant to bridge to this L3 just to trade, so we would have to abstract this away from them (again, wrapped smart contracts may help in the optimistic case where there isn't a conflicting order so they can buy game assets right away).

**Option 2) Stylus**

This option assumes the base chain being Arbitrum.  

> Arbitrum recently introduced a new programming language called Stylus which is much faster & cheaper than EVM. Additionally, it's composable with EVM contracts so you do not run into the same composability tradeoffs as with L3s. However, it does not appear to be able to solve the concurrent issue entirely like the UTxO model. Rather, it might just make the gas cost of failing cheaper. Additionally, it's not live on mainnet at the moment.

Reviewer: I think we can remove this option since it's not immediately useful

**Option 3) Frontend-driven concurrency management**

This option is perhaps the easiest if there is only a single website for the DEX, because the website itself can keep track of orders people are attempting to make and thus avoid conflicting orders being placed. However, this can quickly fall apart if another website appears or if people start making trades by directly interacting with the contract.

> 

Reviewer: I would add a security section that talks about trusting RPCs which is a common assumption in dApps (similar to how we mentioned this in the inverse projection PRC) and how people can run their own node

## Motivation

Many games, due to being data and computation heavy applications, run on sidechains, L2s and appchains as opposed to popular L1 blockchains. This is problematic because liquidity for trading assets live primarily on the L1s (different environments). A common solution to this problem is building a bridge, but bridges have a bad reputation, often require a long delay (especially for optimistic bridges which often require 1 week), and bridging also makes upgrading the game harder as any update to the game state may now also require you to update the data associated with all the bridged state.

Instead of bridging, this standard allows trading the game assets on base chain via creating and filling sell orders in an orderbook DEX smart contract.

## Specification

Every PRC-4 compliant contract must implement the `IOrderbookDex` interface and events:

```solidity
/// @notice Facilitates base-chain trading of an asset that is living on a different app-chain.
/// @dev The contract should never hold any ETH itself.

> interface IOrderbookDex is IERC165 {

Reviewer: We'll need a baseURI specified for this similar to the baseURI format we defined in PRC3 since Tarochi/Paima will have to implement that (as well as a function in the interface that is equivalent to tokenURI)

    struct Order {
        uint256 assetAmount;
        uint256 pricePerAsset;
        bool cancelled;
    }

    event OrderCreated(address indexed seller, uint256 indexed orderId, uint256 assetAmount, uint256 pricePerAsset);

>     event OrderFilled(

Reviewer: We should mention partial fills in the comments somewhere

        address indexed seller,
        uint256 indexed orderId,
        address indexed buyer,
        uint256 assetAmount,
        uint256 pricePerAsset
    );
    event OrderCancelled(address indexed seller, uint256 indexed orderId);

    /// @notice Returns the seller's current `orderId` (index that their new sell order will be mapped to).

>     function getSellerOrderId(address seller) external view returns (uint256);

Reviewer: Should we add this kind of function to the prc3 contract as well? Ideally the functions and structs between these 2 standards look similar in

Author: I don't think it makes too much sense in PRC3, as we only **recommend** the userTokenId to be address-specific incremental counter.

Reviewer: What is the benefit of having this function though? If having this function is actually useful for something, maybe we should make it required for PRC3

Author: I don't see immediate benefits **in our use case**, but it might become useful in the future or for 3rd parties. Better to have this information available than not.

    /// @notice Creates a sell order with incremental seller-specific `orderId` for the specified `assetAmount` at specified `pricePerAsset`.
    /// @dev The order information is saved in a nested mapping `seller address -> orderId -> Order`.
    /// MUST emit `OrderCreated` event.
    function createSellOrder(uint256 assetAmount, uint256 pricePerAsset) external;

    /// @notice Consecutively fills an array of orders identified by the combination `<seller, orderId>`,
    /// by providing an exact amount of ETH and requesting a specific minimum amount of asset to receive.
    /// @dev Transfers portions of msg.value to the orders' sellers according to the price.
    /// The sum of asset amounts of filled orders MUST be at least `minimumAsset`.
    /// If msg.value is more than the sum of orders' prices, it MUST refund the excess back to msg.sender.
    /// An order whose `cancelled` parameter has value `true` MUST NOT be filled.
    /// MUST change the `assetAmount` parameter for the specified order according to how much of it was filled.
    /// MUST emit `OrderFilled` event for each order accordingly.

>     function fillOrdersExactEth(uint256 minimumAsset, address payable[] memory sellers, uint256[] memory orderIds)

Reviewer: We may want a global id for orders we can use as if just need to pass a single id array here it will increase the max number of ids that can be processed in a single tx

unless you had another idea for how to avoid this size limit

Author: Having two arrays instead of one does not pose any threat to hitting the transaction size limit, as realistic scenarios would never come close to the limits (you wouldn't need to be filling 1000 orders at once). I had previously thought about packing the address (20 bytes = uint160) and orderId as uint96 into one uint256, I disregarded that idea because I wanted to keep things simple and it did not really matter storage-wise. But now that you mentioned it, I'm testing for the gas cost difference between these two approaches with regards to **function call**, and I'm seeing around 2-10% gas cost difference. Do you think it is worth making the standard more complicated (considering this probably won't be deployed to expensive eth mainnet)?

Reviewer: If you look at PRC3, we have 2 different IDs (`tokenId` and `userTokenId`). The idea was that `userTokenId` is the wallet-specific one but we can use the global `tokenId` wherever we don't want to bother passing a `<userTokenId, address>` pair

Author: Yes but that's because PRC3 is IERC721, so we **had to** have global tokenId to conform with the IERC721 in functions like `tokenURI(uint256 tokenId)`. We don't need that here, it would just be a waste of storage - If we want to have the single id, we can simply construct it by packing seller address and seller-specific order id into one uint256 id.

    /// @notice Consecutively fills an array of orders identified by the combination `<seller, orderId>`,
    /// by providing a possibly surplus amount of ETH and requesting an exact amount of asset to receive.
    /// @dev Transfers portions of msg.value to the orders' sellers according to the price.
    /// The sum of asset amounts of filled orders MUST be exactly `assetAmount`. Excess ETH MUST be returned back to `msg.sender`.
    /// An order whose `cancelled` parameter has value `true` MUST NOT be filled.
    /// MUST change the `assetAmount` parameter for the specified order according to how much of it was filled.
    /// MUST emit `OrderFilled` event for each order accordingly.
    /// If msg.value is more than the sum of orders' prices, it MUST refund the difference back to msg.sender.
    function fillOrdersExactAsset(uint256 assetAmount, address payable[] memory sellers, uint256[] memory orderIds)
        external
        payable;

    /// @notice Cancels the sell order identified by combination `<msg.sender, orderId>`, making it unfillable.
    /// @dev MUST change the `cancelled` parameter for the specified order to `true`.
    /// MUST emit `OrderCancelled` event.

>     function cancelSellOrder(uint256 orderId) external;

Reviewer: From a gane perspective, when you see a cancel event emitted, how do you know how much gold to unlock for which user?

orderId here would either have to be a gobal id so we can look it up or we need more context

Author: This is the cancel function, the event looks like this: `event OrderCancelled(address indexed seller, uint256 indexed orderId);` so it does indeed have the seller part. It just is not needed in the function since you should only be able to cancel your own orders. As is described above, "Cancels the sell order identified by combination `<msg.sender, orderId>`"

       image?: string;          // OPTIONAL Asset URL Image (1:1 200px Image)
       totalSupply: number;     // Total number of assets
   }
   ```

3. Get ERC1155 tokens of `asset` for the specified `wallet` that have been minted or owned in a valid way. This is used by the DEX to get the list of valid assets that user is able to create sell orders for.

   `GET /dex/{asset}/wallet/{wallet}`

   - asset: valid name for specific game asset token.
   - wallet: wallet to query for asset

   RESPONSE

   ```js
   {
     total: number; // Total number of assets owned
     stats: {
       tokenId: number; // ERC1155 Token ID
       amount: number; // Number of assets owned
     }
     [];
   }
   ```

4. Gets valid created Sell Orders. This is used by the DEX to get the list of valid Sell Orders to display to users wanting to buy. Ordered by lowest price.

   `GET /dex/{asset}/orders?seller=wallet&page=number&limit=number`

   - asset: valid name for specific game asset token.
   - seller (OPTIONAL): fetch where wallet address matches wallet
   - page (OPTIONAL): results page number, default = 1
   - limit (OPTIONAL): 10, 25, 50, 100, default = 25

   RESPONSE

   ```js

>    {
>      stats: {
>        orderId: number; // Order unique ID
>        seller: string; // Seller wallet
>        tokenId: number; // ERC1155 TokenID
>        amount: number; // Number of assets for sale
>        price: string; // Price per asset
>      }
>      [];
>    }

Reviewer 2: ```suggestion
   {
     stats: {
       orderId: number; // Order unique ID
       seller: string; // Seller wallet
       tokenId: number; // ERC1155 TokenID
       amount: number; // Number of assets for sale
       price: string; // Price per asset
       asset: string; // Asset address
       makerFee: number; // Maker Fee 
       takerFee: number; // Taker Fee
     }
     [];
   }
```

    ) external payable;

    /// @notice Cancels the sell order of `asset` with asset-specific `orderId`, transferring the order's assets back to the seller.
    /// @dev MUST revert if the order's seller is not `msg.sender`.
    /// MUST change the `assetAmount` parameter for the specified order to `0`.
    /// MUST emit `OrderCancelled` event.
    /// MUST transfer the order's `assetAmount` of `asset` with `assetId` back to the seller.
    function cancelSellOrder(address asset, uint256 orderId) external;

    /// @notice Cancels a batch of sell orders of `asset` with asset-specific `orderIds`, transferring the orders' assets back to the seller.
    /// @dev This is a batched version of `cancelSellOrder` that simply iterates through the array to call said function.
    function cancelBatchSellOrder(address asset, uint256[] memory orderIds) external;

    /// @notice Withdraws the contract balance (containing collected fees) to the owner. Executable only by the owner.
    /// @dev MUST transfer the entire contract balance to the owner.
    /// MUST revert if called by unauthorized account.
    /// MUST emit `FeesWithdrawn` event.
    function withdrawFees() external;

> }

Reviewer: Maybe we need to add the claim function and the rationale for why it's needed? (the rationale thread we had on Slack)

Author: Added!
